Implement the Mail System

4Benj opened the pull request:
Changes:
- Get mail protos
- Send mail command
- Database support for mail and helper functions in Player class
- DatabaseHelper.GetAllPlayers (I'm not actually sure if this works and testing is required)

Things to be complete later:
- Mail template system (Saving a template to a database and send it using the `/sendmail` command. The `/sendmail` command already has a place for it to be easily implemented on [line 62](https://github.com/Grasscutters/Grasscutter/pull/115/files#diff-7acd01e31d9a50857d786a6c7b6e8b09be1833d0d7a264b2b1c6745fb7088c5bR62).
- Gift mail box. Currently it is completely disabled, it should be relatively simple to add by slightly modifying `GetAllMailRsp`. However, the amount of issues I have getting that working I'm kind of exhausting working on mail atm lol.

4Benj committed:
Update protos

4Benj committed:
Update protos

4Benj committed:
Initial work on mail

4Benj committed:
Initial work on mail

4Benj committed:
Added GetMailItemReq/Rsp & PacketMailChangeNotify

Mail still doesn't work for whatever reason. I also added a temporary SendMailCommand just to receive PacketMailChangeNotify.
I feel like I'm about to lose my mind.

4Benj committed:
Added GetMailItemReq/Rsp & PacketMailChangeNotify

Mail still doesn't work for whatever reason. I also added a temporary SendMailCommand just to receive PacketMailChangeNotify.
I feel like I'm about to lose my mind.

BaiSugar commented:
try GetAllMailReq ,GetAllMailRsp,MailChangeNotify?

lilmayofuksu commented:
![image](https://user-images.githubusercontent.com/11977128/164720209-335c4fb0-23c3-4068-a06d-c2384f83ce75.png)

It seems that server needs to reply to the GetAllMailReq packet

4Benj committed:
Added database support for mail

Send mail command now works from console (to use it from console change the uuid in the 'SendMailCommand' file

4Benj committed:
Added database support for mail

Send mail command now works from console (to use it from console change the uuid in the 'SendMailCommand' file

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

4Benj commented:
I've made a few updates and now the current issue is that the client seems to receive the mail and the mail icon is the menu has a badge indicating there is unread mail. However, there is nothing in the mail box when you check it...

It seems that it only updates when the client receives `MailChangeNotify` which doesn't make sense as that doesn't seem to be the same as the official servers. It's possible that the client receives `GetAllMailRsp` too early but if not I may have to send `MailChangeNotify` after sending `GetAllMailRsp`.

PasserbyAlpha commented:
I've got some Exceptions when loading the game, and I guess it's because of the version of java (development branch has been switched to java 16 from 1.8). Add @Entity before the class definition of Mail, MailContent and MailItem will solve (referring to https://github.com/Grasscutters/Grasscutter/issues/140#issuecomment-1107388851)

Besides, according to the message of MailChangeNotify, I doubt that maybe isRead and isAttachmentGot shouldn't appear in MailChangeNotify at all, instead of appearing with value false. At least those key don't appear in the message that I sniffed. I guess that client will only add MailChangeNotify without those key to original mail list. However, I'm not familiar with those libs and don't know how to make it. The only thing I tried is that we cannot just change boolean to Boolean and set null to them (or there will be exceptions)

4Benj commented:
> I've got some Exceptions when loading the game, and I guess it's because of the version of java (development branch has been switched to java 16 from 1.8). Add `@Entity` before the class definition of Mail, MailContent and MailItem will solve it (referring to https://github.com/Grasscutters/Grasscutter/issues/140#issuecomment-1107388851)

I kinda merged my project into the latest with realising that there going to huge changes like that and didn't actually Brest after merging because I was tired.

> Besides, according to the message of MailChangeNotify, I doubt that maybe isRead and isAttachmentGot shouldn't appear in MailChangeNotify at all, instead of appearing with value false. At least those key don't appear in the message that I sniffed. I guess that client will only add MailChangeNotify without those key to original mail list. However, I'm not familiar with those libs and don't know how to make it. The only thing I learned is that we cannot just change boolean to Boolean and set null to them (or there will be exceptions)

Thanks for that, I've been unable to sniff the packets myself for some particular reason, so I've had to reconstruct some things without a template and have just been interpreting some things based on their protos (which has worked out so far lol). When I get the chance I can remove isRead and isAttachmentGot from the packet being sent from the client later.

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

4Benj committed:
Java-16 compatiable mail system (whoops)

4Benj committed:
Java-16 compatiable mail system (whoops)

4Benj committed:
THE MAILBOX ACTUALLY WORKS CORRECTLY NOW

4Benj committed:
THE MAILBOX ACTUALLY WORKS CORRECTLY NOW

4Benj commented:
Big thanks to @lilmayofuksu for his help on the discord. 
This fucking game sometimes.
I had to send a "empty" packet notifying the game the game the mailbox was empty or it would fucking break.
I probably wouldn't of been able to do this without you so big thanks lol.

4Benj commented:
Now to add the missing packets and it'll be all done 😆 
See you after my next journey to the underworld :)

4Benj committed:
Receiving items in mail

4Benj committed:
Receiving items in mail

4Benj committed:
Deleting mail + a bit of cleanup

4Benj committed:
Deleting mail + a bit of cleanup

4Benj committed:
Mail read notify

4Benj committed:
Mail read notify

4Benj committed:
Added ChangeMailStarNotify (change importance level)

4Benj committed:
Added ChangeMailStarNotify (change importance level)

4Benj committed:
Weird git stuff

4Benj committed:
More weird git stuff

4Benj committed:
Forgot to implement levels into GetMailItemRsp

Also added promoteLevels

4Benj committed:
Send Mail Command!

Almost done! Wooooo!

4Benj commented:
My madness is finally about to come to an end but I have some more testing to do but I'll do that tomorrow after school and after I get a bit of homework done (I've been putting it off for GC lol).
If anyone wants to test and encounter any issues, please post a comment here so I can fix it (Or you can submit a PR to my repo)!

4Benj committed:
Improvements and send all mail fix

Mail ID no longer exists and is handled by it's position in the mail array

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

4Benj committed:
Didn't mean to push the handbook

4Benj commented:
The mail system seems to be complete. There are further additions that can be added. However, the system works and the changes are minor and not required.

4Benj commented:
No matter what I added @KingRainbow44 always seems to have to review it :joy:

Konoyoru435 commented on src/main/java/emu/grasscutter/database/DatabaseHelper.java:
```diff
@@ -90,6 +90,10 @@ public static boolean deleteAccount(String username) {
 	}
 
+	public static List<GenshinPlayer> getAllPlayers() {
+		return DatabaseManager.getDatastore().find(GenshinPlayer.class).stream().toList();
```
Thanks for this line of code, it helped me a lot, I was converting wetabq's kt to java, and there was a `toList`, and I found here that I could use `stream`

KingRainbow44 commented on src/main/java/emu/grasscutter/command/commands/SendMailCommand.java:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        if (!mailBeingConstructed.containsKey(senderId)) {
+            switch (args.size()) {
```
update switch case to `case "foo" -> bar()` pls

KingRainbow44 commented on src/main/java/emu/grasscutter/command/commands/SendMailCommand.java:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    public String getConstructionArgs(int stage) {
```
private method maybe?

KingRainbow44 reviewed (approved):
looks good

WetABQ commented:
after https://github.com/Grasscutters/Grasscutter/pull/226, u should del those Outer classes, which are generated by protoc
try to pull and merge, to resolve merge conflicts of u pr, thanks very much

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

Melledy commented:
Waiting on your conflict merge.

4Benj committed:
Removed debug messages, readded protos, fixed gradle

4Benj commented:
> Waiting on your conflict merge.

All done :)

4Benj committed:
Merge remote-tracking branch 'upstream/development' into dev-mail

4Benj committed:
GenshinPlayer fix (for two issues) and updated sendmail with Magix's suggestions

4Benj commented:
Fixed the build issues and implemented @KingRainbow44 suggested changes
